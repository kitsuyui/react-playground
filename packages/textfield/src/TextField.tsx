import {
  ComponentPropsWithoutRef,
  forwardRef,
  useCallback,
  useEffect,
  useRef,
  useState,
} from 'react'
import React from 'react'

import { useCombinedRefs } from './utils'

type WrappedProps = ComponentPropsWithoutRef<'input'>

type alternateProps = {
  onInputChunk?: (value: string) => void
  onChangeInputting?: (inputting: boolean) => void
  value?: string
}

type excludeProps =
  | 'onChange'
  | 'onCompositionStart'
  | 'onCompositionUpdate'
  | 'onCompositionEnd'
  | 'value'
  | 'type'

export type WrapperProps = Omit<WrappedProps, excludeProps> & alternateProps

export const TextField = forwardRef<HTMLInputElement, WrapperProps>(
  (props, ref) => {
    const innerRef = useRef<HTMLInputElement>(null)
    const { onInputChunk, onChangeInputting } = props
    const [internalValue, setInternalValue] = useState(props.value ?? '')
    const [isInputting, setIsInputting] = useState(false)
    const combinedRef = useCombinedRefs(innerRef, ref)

    // Use Object.assign({}, props) instead of just { ...props } because it must create deep copy.
    const {
      onInputChunk: _1,
      onChangeInputting: _2,
      ...propsExcludedWrapperProps
    } = Object.assign({}, props)
    const onBlur = propsExcludedWrapperProps.onBlur

    useEffect(() => {
      setInternalValue(props.value ?? '')
    }, [props.value])

    const handleChange = useCallback(() => {
      const text = innerRef?.current?.value ?? ''
      setInternalValue(text)
      if (isInputting) return
      onInputChunk?.(text)
    }, [isInputting, onInputChunk])

    const handleCompositionChange = useCallback(() => {
      const text = innerRef?.current?.value ?? ''
      setInternalValue(text)
      setIsInputting(true)
      onChangeInputting?.(true)
    }, [onChangeInputting])

    const handleCompositionEnd = useCallback(() => {
      const text = innerRef?.current?.value ?? ''
      setInternalValue(text)
      onChangeInputting?.(false)
      onInputChunk?.(text)
      setIsInputting(false)
    }, [onInputChunk, onChangeInputting])

    const handleOnBlur = useCallback(
      (e: React.FocusEvent<HTMLInputElement>) => {
        const text = innerRef?.current?.value ?? ''
        setInternalValue(text)
        setIsInputting(false)
        onChangeInputting?.(false)
        onInputChunk?.(text)
        onBlur?.(e)
      },
      [onInputChunk, onChangeInputting, onBlur]
    )

    return (
      <input
        {...propsExcludedWrapperProps}
        type="text"
        ref={combinedRef}
        value={internalValue}
        onCompositionStart={handleCompositionChange}
        onCompositionUpdate={handleCompositionChange}
        onCompositionEnd={handleCompositionEnd}
        onBlur={handleOnBlur}
        onChange={handleChange}
      />
    )
  }
)

TextField.displayName = 'TextField'
