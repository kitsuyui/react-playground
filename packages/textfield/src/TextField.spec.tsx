import { expect, it, describe, vi, beforeEach } from 'vitest'

import { render, cleanup, fireEvent } from '@testing-library/react'
import userEvent from '@testing-library/user-event'

import { TextField, type TextFieldRef } from './TextField'
import React from 'react'

const ExampleTextField = (props: {
  initialText?: string
  onChange?: (value: string) => void
  onChangeInputting?: (inputting: boolean) => void
}) => {
  const { initialText, onChange, onChangeInputting } = props
  const ref = React.useRef<TextFieldRef>(null)
  const [text, setText] = React.useState(initialText ?? '')
  return (
    <>
    <TextField
      value={text}
      ref={ref}
      onInputChunk={(chunk) => {
        setText(chunk)
        onChange?.(chunk)
      }}
      onChangeInputting={(inputting) => {
        onChangeInputting?.(inputting)
      }}
    />
    <button
      type="button"
      onClick={() => {
        ref.current?.clear()
      }}
    >
      Clear
    </button>
    </>
  )
}

beforeEach(() => {
  cleanup()
})

describe('TextField', () => {
  it('should render', () => {
    const { asFragment } = render(
      <TextField value="Hello, World!" onInputChunk={() => {}} />
    )
    expect(asFragment()).toMatchSnapshot()
  })

  it('treats empty string when value is not provided', () => {
    const { asFragment } = render(
      <TextField/>
    )
    expect(asFragment()).toMatchSnapshot()
    const textfield = document.querySelector('input') as HTMLInputElement
    expect(textfield).toBeInstanceOf(HTMLInputElement)
    expect(textfield.value).toBe('')
  })

  it('should handle input', async () => {
    const handleInputChunk = vi.fn()
    const rendered = render(
      <ExampleTextField
        initialText=""
        onChange={handleInputChunk}
      />
    )
    expect(rendered.asFragment()).toMatchSnapshot()
    const TextField = rendered.getByRole('textbox') as HTMLInputElement
    expect(TextField).toBeInstanceOf(HTMLInputElement)
    expect(TextField.value).toBe('')
    await userEvent.click(TextField) // focus the TextField
    await userEvent.type(TextField, 'Hello, World!')
    expect(TextField.value).toBe('Hello, World!')
    expect(rendered.asFragment()).toMatchSnapshot()
  })

  it('should handle input with composition change events', async () => {
    const handleInputChunk = vi.fn()
    const handleInputting = vi.fn()
    const rendered = render(
      <ExampleTextField
        initialText=""
        onChange={handleInputChunk}
        onChangeInputting={handleInputting}
      />
    )
    expect(rendered.asFragment()).toMatchSnapshot()
    const TextField = rendered.getByRole('textbox') as HTMLInputElement
    expect(TextField).toBeInstanceOf(HTMLInputElement)
    expect(TextField.value).toBe('')
    await userEvent.click(TextField) // focus the TextField
    // userEvent.type does not support composition events directly,
    // so we simulate composition events manually
    expect(handleInputting).toHaveBeenCalledTimes(0)
    fireEvent.compositionStart(TextField)
    expect(handleInputting).toHaveBeenCalledTimes(1)
    fireEvent.change(TextField, { target: { value: 'Hello' } })
    expect(handleInputting).toHaveBeenCalledTimes(1)
    // When composition is in progress, TextField value is changed but not called onInputChunk
    expect(TextField.value).toBe('Hello')

    // After composition ends, the value is committed
    fireEvent.compositionEnd(TextField)
    expect(handleInputting).toHaveBeenCalledTimes(2)
    expect(TextField.value).toBe('Hello')
    expect(handleInputChunk).toHaveBeenCalledTimes(1)
    expect(handleInputChunk).toHaveBeenCalledWith('Hello')
  })

  it('clear should reset value', async () => {
    const handleInputChunk = vi.fn()
    const rendered = render(
      <ExampleTextField
        initialText="Hello, World!"
        onChange={handleInputChunk}
      />
    )
    expect(rendered.asFragment()).toMatchSnapshot()
    const TextField = rendered.getByRole('textbox') as HTMLInputElement
    expect(TextField).toBeInstanceOf(HTMLInputElement)
    expect(TextField.value).toBe('Hello, World!')
    await userEvent.click(TextField) // focus the TextField
    await userEvent.type(TextField, '2')
    expect(TextField.value).toBe('Hello, World!2')
    expect(rendered.asFragment()).toMatchSnapshot()
    const clearButton = rendered.queryByText('Clear')
    expect(clearButton).toBeInstanceOf(HTMLButtonElement)
    expect(clearButton).not.toBeNull()
    await userEvent.click(clearButton as HTMLButtonElement)
    expect(TextField.value).toBe('')
    expect(rendered.asFragment()).toMatchSnapshot()
  })
})
