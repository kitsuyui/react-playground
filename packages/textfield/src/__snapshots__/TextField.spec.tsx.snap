// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TextField > clear should reset value 1`] = `
<DocumentFragment>
  <input
    type="text"
    value="Hello, World!"
  />
  <button
    type="button"
  >
    Clear
  </button>
</DocumentFragment>
`;

exports[`TextField > clear should reset value 2`] = `
<DocumentFragment>
  <input
    type="text"
    value="Hello, World!2"
  />
  <button
    type="button"
  >
    Clear
  </button>
</DocumentFragment>
`;

exports[`TextField > clear should reset value 3`] = `
<DocumentFragment>
  <input
    type="text"
    value=""
  />
  <button
    type="button"
  >
    Clear
  </button>
</DocumentFragment>
`;

exports[`TextField > should handle input 1`] = `
<DocumentFragment>
  <input
    type="text"
    value=""
  />
  <button
    type="button"
  >
    Clear
  </button>
</DocumentFragment>
`;

exports[`TextField > should handle input 2`] = `
<DocumentFragment>
  <input
    type="text"
    value="Hello, World!"
  />
  <button
    type="button"
  >
    Clear
  </button>
</DocumentFragment>
`;

exports[`TextField > should handle input with composition change events 1`] = `
<DocumentFragment>
  <input
    type="text"
    value=""
  />
  <button
    type="button"
  >
    Clear
  </button>
</DocumentFragment>
`;

exports[`TextField > should render 1`] = `
<DocumentFragment>
  <input
    type="text"
    value="Hello, World!"
  />
</DocumentFragment>
`;
